src/dev/platform/f3xx/f302x8/Timerf302x8.cpp

```diff
@@ -0,0 +1,168 @@
+    uint8_t interruptIdx = getTimerInterruptIndex(htim->Instance);
+    if (timerInterruptHandlers[interruptIdx] != nullptr) {
+        timerInterruptHandlers[interruptIdx](htim);
+    }
+}
+
+/**
+ * Returns the corresponding index for the given timer device as stored in the
+ * timerInterruptHandlers array
+ * @param htim the TIM_TypeDef pointer to the timer device
+ * @return an uint8_t between 0-3 corresponding to an element of timerInterruptHandlers
+ */
+uint8_t getTimerInterruptIndex(TIM_TypeDef *peripheral) {
+    uint8_t interruptIdx;
+
+    if (peripheral == TIM2) {
+        interruptIdx = static_cast<uint8_t>(timerInterruptIndex::TIM2_IDX);
+    } else if (peripheral == TIM15) {
+        interruptIdx = static_cast<uint8_t>(timerInterruptIndex::TIM15_IDX);
+    } else if (peripheral == TIM16) {
+        interruptIdx = static_cast<uint8_t>(timerInterruptIndex::TIM16_IDX);
+    } else if (peripheral == TIM17) {
+        interruptIdx = static_cast<uint8_t>(timerInterruptIndex::TIM17_IDX);
+    } else {
+        interruptIdx = -1;
+    }
+
+    return interruptIdx;
+}
+
+namespace EVT::core::DEV {
+
+Timerf302x8::Timerf302x8(TIM_TypeDef *timerPeripheral, uint32_t clockPeriod,
+                         void (*irqHandler)(TIM_HandleTypeDef *htim)) {
+    timerInterruptHandlers[getTimerInterruptIndex(timerPeripheral)] = irqHandler;
+    initTimer(timerPeripheral, clockPeriod);
+    this->halTimer = &halTimers[getTimerInterruptIndex(timerPeripheral)];
+}
+
+void Timerf302x8::initTimer(TIM_TypeDef *timerPeripheral, uint32_t clockPeriod) {
+    this->clockPeriod = clockPeriod;
+    auto& htim = halTimers[getTimerInterruptIndex(timerPeripheral)];
+
+    htim.Instance = timerPeripheral;
+    uint32_t prescaler = EVT::core::platform::CLK_SPEED / 1000;
+    htim.Init.Prescaler = prescaler;    // Sets f_CK_PSC to 1000 Hz
+    // Allows period increments of 1 ms with max of 2^(32) ms.
+    htim.Init.CounterMode = TIM_COUNTERMODE_UP;
+    htim.Init.Period = clockPeriod;
+    htim.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+    htim.Init.AutoReloadPreload = TIM_AUTORELOAD_PRELOAD_ENABLE;
+    HAL_TIM_Base_Init(&htim);
+
+    TIM_ClockConfigTypeDef clockConfig = {0};
+    TIM_MasterConfigTypeDef masterConfig = {0};
+
+    clockConfig.ClockSource = TIM_CLOCKSOURCE_INTERNAL;
+    HAL_TIM_ConfigClockSource(&htim, &clockConfig);
+
+    masterConfig.MasterOutputTrigger = TIM_TRGO_RESET;
+    masterConfig.MasterSlaveMode = TIM_MASTERSLAVEMODE_DISABLE;
+    HAL_TIMEx_MasterConfigSynchronization(&htim, &masterConfig);
+
+    HAL_TIM_Base_Start_IT(&htim);
+}
+
+
+void Timerf302x8::startTimer(void (*irqHandler)(TIM_HandleTypeDef *htim)) {
+    TIM_TypeDef *timerPeripheral = this->halTimer->Instance;
+    stopTimer();
+    timerInterruptHandlers[getTimerInterruptIndex(timerPeripheral)] = irqHandler;
+    initTimer(timerPeripheral, this->clockPeriod);
+}
+
+void Timerf302x8::stopTimer() {
+    HAL_TIM_Base_DeInit(this->halTimer);
```

Comment:
Do we want to completely de-init the timer here?  Another function call is `HAL_TIM_Base_Stop_IT()` which just pauses the timer module.  Realistically the functionality appears to be the same either way.

Reply:
We don't have to complete de-init. I imagine this will be used like a "pause" for the timer.

Reply:
I agree.